pyteal/ast/subroutine.py

```diff
@@ -1006,7 +1007,19 @@ def __call__(self, subroutine: SubroutineDefinition) -> SubroutineDeclaration:
             abi_output_kwargs[output_kwarg_info.name] = output_carrying_abi
 
         # Arg usage "B" supplied to build an AST from the user-defined PyTEAL function:
-        subroutine_body = subroutine.implementation(*loaded_args, **abi_output_kwargs)
+        subroutine_body: Expr
+        if not self.use_frame_pt:
+            subroutine_body = subroutine.implementation(
+                *loaded_args, **abi_output_kwargs
+            )
+        else:
+            prev_proto: Optional[Proto] = SubroutineEval.current_proto
+            SubroutineEval.current_proto = proto
```

Comment:
Now that you have followed my suggestion and removed the python context-manager I see better the original purpose. We need a way to transmit to ABI Types that they are in a particular proto context and then use the frame opcodes accordingly, instead of scratch var's. Since we want the ABI Types to detect the proto context without explicitly being told about this, we need to pass this information through a global (or static) variable. Without a python context-manager, we are left to directly manipulate the static variable before and after subroutine evaluation, as is happening here. However, this manual manipulation is exactly what context-managers are meant to relieve us from.
So I must backtrack from my original suggestion and suggest that context managers be employed. I would accept reverting to the original code.
However, my original remarks about context managers being confusing still stand. So I wonder if there is a way to make the intent simpler. Perhaps it's just a matter of variable naming.

The original usage pattern looked like:
```python
with SubroutineEval.Context.CompileWithFrameContext(proto):
    subroutine_body = subroutine.implementation(...)
```

Perhaps the following would be easier to grok?

```python
with frame_pointer_context(proto):
    ...
```
The implementation might look like the following (which I basically cribbed from [this RealPython article](https://realpython.com/python-with-statement/#creating-function-based-context-managers) ) but you could also using a class-based context and keep the running `proto` as a static variable, instead of the global that I'm using:

```python
from contextlib import contextmanager

_fp_context_proto: Optional[Proto] = None

@contextmanager
def frame_pointer_context(proto: Proto):
    global _fp_context_proto
    tmp, _fp_context_proto = _fp_context_proto, proto
    yield proto # might as well yield the proto
    _fp_context_proto = tmp
```


But again, reverting to the original implementation is good 'nuf. 

And apologies for asking you to revert the exact thing I previously asked you to do.